### PR TITLE
[SPARK-41968][CORE][SQL] Refactor `ProtobufSerDe` to  `ProtobufSerDe[T]`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims refator `ProtobufSerDe` to  `ProtobufSerDe[T]`, the main change of `ProtobufSerDe` as follows:

- Change the definition of `ProtobufSerDe` to `ProtobufSerDe[T]`
- Remove `supportClass` method from `ProtobufSerDe[T]` and use reflection in `KVStoreProtobufSerializer` to obtain the actual type of `T` as `serializerMap` key
- Change the input parameter type of `serialize` function from `Any` to `T`
- Change the return value type of `deserialize` function method from `Any` to `T`

Then, all the subclasses of `ProtobufSerDe[T]` are refactored and code cleaned in this pr.

### Why are the changes needed?
Refactor `ProtobufSerDe` and code cleanup.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass Github Actions

### CODE REVIEW

**Comment by LuciferYang** [comment]
GA passed

File: core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala [hunk 1]
```diff
@@ -26,7 +27,7 @@ import org.apache.spark.status.KVUtils.KVStoreScalaSerializer
 private[spark] class KVStoreProtobufSerializer extends KVStoreScalaSerializer {
   override def serialize(o: Object): Array[Byte] =
     KVStoreProtobufSerializer.getSerializer(o.getClass) match {
-      case Some(serializer) => serializer.serialize(o)
+      case Some(serializer) => serializer.serialize(o.asInstanceOf[Any])
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala:30]: Do we really need call `asInstanceOf[Any]`?

**Comment by gengliangwang** [comment]
@LuciferYang Thanks for the work. I will take a closer look tomorrow.

File: core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala (hunk 1 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala]: shouldn't be necessary, remove it in [2568fc0](https://github.com/apache/spark/pull/39487/commits/2568fc0d654ed13c04bd335ece01503d3d5e0c71), waiting CI

File: core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala [hunk 2]
```diff
@@ -40,10 +41,16 @@ private[spark] class KVStoreProtobufSerializer extends KVStoreScalaSerializer {
 
 private[spark] object KVStoreProtobufSerializer {
 
- private[this] lazy val serializerMap: Map[Class[_], ProtobufSerDe] =
-   ServiceLoader.load(classOf[ProtobufSerDe])
-     .asScala.map(serDe => serDe.supportClass -> serDe).toMap
+  private[this] lazy val serializerMap: Map[Class[_], ProtobufSerDe[Any]] = {
+    def getGenericsType(klass: Class[_]): Class[_] = {
+      klass.getGenericInterfaces.head.asInstanceOf[ParameterizedType]
+        .getActualTypeArguments.head.asInstanceOf[Class[_]]
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala:47]:
```suggestion
        .getRawType.asInstanceOf[Class[_]]
```

Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala:47]: This is neat

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala:47]: I want to double check, do you want to change it to

```
klass.getGenericInterfaces.head.asInstanceOf[ParameterizedType].getRawType.asInstanceOf[Class[_]]
```

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala:47]: The result of 
```
klass.getGenericInterfaces.head.asInstanceOf[ParameterizedType]
  .getRawType.asInstanceOf[Class[_]]
```
is always `org.apache.spark.status.protobuf.ProtobufSerDe`

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala:47]: The result of 
```
klass.getGenericInterfaces.head.asInstanceOf[ParameterizedType]
  .getActualTypeArguments.head.asInstanceOf[Class[_]]
```
are 

```
org.apache.spark.status.RDDStorageInfoWrapper
org.apache.spark.status.ApplicationInfoWrapper
org.apache.spark.status.ApplicationEnvironmentInfoWrapper
org.apache.spark.status.CachedQuantile
org.apache.spark.status.ExecutorStageSummaryWrapper
org.apache.spark.status.StreamBlockData
org.apache.spark.status.TaskDataWrapper
org.apache.spark.status.JobDataWrapper
org.apache.spark.status.ResourceProfileWrapper
org.apache.spark.status.SpeculationStageSummaryWrapper
org.apache.spark.status.ExecutorSummaryWrapper
org.apache.spark.status.ProcessSummaryWrapper
org.apache.spark.status.RDDOperationGraphWrapper
org.apache.spark.status.StageDataWrapper
```
This is what we need

Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/status/protobuf/KVStoreProtobufSerializer.scala:47]: You are right.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
@LuciferYang Thanks for the work.
Merging to master

**Comment by LuciferYang** [comment]
Thanks @gengliangwang

**Pull request closed without merging**